pkg/registry/flowcontrol/ensurer/strategy.go

```diff
@@ -0,0 +1,303 @@
+	value, _ := accessor.GetAnnotations()[flowcontrolv1beta1.AutoUpdateAnnotationKey]
+	if autoUpdate, err := strconv.ParseBool(value); err == nil {
+		return autoUpdate
+	}
+
+	// We are here because of either a or b:
+	// a. the annotation key is missing.
+	// b. the annotation key is present but the value does not represent a boolean.
+	// In either case, if the operator hasn't changed the spec, we can safely auto update.
+	// Please note that we can't protect the changes made by the operator in the following scenario:
+	// - The operator deletes and recreates the same object with a variant spec (generation resets to 1).
+	if accessor.GetGeneration() == 1 {
+		return true
+	}
+	return false
+}
+
+// shouldUpdateAnnotation determines whether the current value of the auto-update annotation
+// key matches the desired value.
+func shouldUpdateAnnotation(accessor metav1.Object, desired bool) bool {
+	if value, ok := accessor.GetAnnotations()[flowcontrolv1beta1.AutoUpdateAnnotationKey]; ok {
+		if current, err := strconv.ParseBool(value); err == nil && current == desired {
+			return false
+		}
+	}
+
+	return true
+}
+
+// setAutoUpdateAnnotation sets the auto-update annotation key to the specified value.
+func setAutoUpdateAnnotation(accessor metav1.Object, autoUpdate bool) {
+	if accessor.GetAnnotations() == nil {
+		accessor.SetAnnotations(map[string]string{})
+	}
+
+	accessor.GetAnnotations()[flowcontrolv1beta1.AutoUpdateAnnotationKey] = strconv.FormatBool(autoUpdate)
+}
+
+// ensureConfiguration ensures the boostrap configurationWrapper on the cluster based on the specified strategy.
+func ensureConfiguration(wrapper configurationWrapper, strategy ensureStrategy, bootstrap configurationObject) error {
+	name := bootstrap.GetName()
+	configurationType := strategy.Name()
+
+	current, err := wrapper.Get(bootstrap.GetName())
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to retrieve %s type=%s name=%q error=%w", wrapper.TypeName(), configurationType, name, err)
+		}
+
+		if strategy.ShouldCreate() {
+			if _, err := wrapper.Create(bootstrap); err != nil {
+				return fmt.Errorf("cannot create %s type=%s name=%q error=%w", wrapper.TypeName(), configurationType, name, err)
+			}
+
+			klog.V(2).InfoS(fmt.Sprintf("Successfully created %s", wrapper.TypeName()), "type", configurationType, "name", name)
+			return nil
+		}
+
+		klog.V(3).InfoS(fmt.Sprintf("Skipping creation of %s", wrapper.TypeName()), "type", configurationType, "name", name)
+		return nil
+	}
+
+	klog.V(5).InfoS(fmt.Sprintf("The %s already exists, checking whether it is up to date", wrapper.TypeName()), "type", configurationType, "name", name)
+	newObject, update, err := strategy.ShouldUpdate(current, bootstrap)
+	if err != nil {
+		return fmt.Errorf("failed to determine whether auto-update is required for %s type=%s name=%q error=%w", wrapper.TypeName(), configurationType, name, err)
+	}
+	if !update {
+		if klog.V(3).Enabled() {
```

Comment:
Should this be logged at --v=3 - it might be a bit spammy imho.

Reply:
changed it to `--v=5`